Rebuild/Rebase for Linux 5.1

I've not made any updates recently, but it shouldn't be too hard to get working again upon 5.1[.x].

Changing `security/Kconfig` + `security/Makefile` was simple and mechanical.

Building `can-exec` fails:

> security/can-exec/can_exec_lsm.c:239:1: error: type defaults to ‘int’ in declaration of ‘security_initcall’ [-Werror=implicit-int]
> security/can-exec/can_exec_lsm.c:239:1: warning: parameter names (without types) in function declaration
> security/can-exec/can_exec_lsm.c:227:19: warning: ‘can_exec_init’ defined but not used [-Wunused-function]

Looks like this:

> security_initcall(can_exec_init);

Should be come this:

>  DEFINE_LSM(can_exec_init) = {
>        .init = can_exec_init,
>        .name = "can-exec",
>};

Recompiling now to test..

--- a/.gitignore
+++ b/.gitignore
@@ -1,0 +1,4 @@
+*.a
+*.o
+*.builtin
+*.order

--- a/README.md
+++ b/README.md
@@ -24,12 +24,11 @@ and they are listed below (in order oldest to most recent):
 
 ## Installation
 
-The code has been tested upon kernels as recent as 4.17.8.
+The code has been tested upon kernels as recent as 5.1
 
-Copy the contents of `security/` into your local Kernel-tree, and run
-`make menuconfig` to enable the appropriate options.
+Copy the contents of `security/` into your local Kernel-tree, and run `make menuconfig` to enable the appropriate options.
 
-**NOTE**: Over time the two files `security/Kconfig` & `security/Makefile` might need resyncing from master - but you can look for mentions of `CAN_EXEC`, `HASH_CHECK`, & `WHITELIST` to see what I've done to add the modules.
+**NOTE**: Over time the two files `security/Kconfig` & `security/Makefile` might need resyncing with the base versions installed with the Linux source-tree, you can look for mentions of `CAN_EXEC`, `HASH_CHECK`, & `WHITELIST` to see what I've done to add the modules.
 
 For a Debian GNU/Linux host, building a recent kernel, these are the dependencies you'll need to install:
 

--- a/security/Kconfig
+++ b/security/Kconfig
@@ -4,7 +4,7 @@
 
 menu "Security options"
 
-source security/keys/Kconfig
+source "security/keys/Kconfig"
 
 config SECURITY_DMESG_RESTRICT
 	bool "Restrict unprivileged access to the kernel syslog"
@@ -40,8 +40,7 @@ config SECURITYFS
 	bool "Enable the securityfs filesystem"
 	help
 	  This will build the securityfs filesystem.  It is currently used by
-	  the TPM bios character driver and IMA, an integrity provider.  It is
-	  not used by SELinux or SMACK.
+	  various security modules (AppArmor, IMA, SafeSetID, TOMOYO, TPM).
 
 	  If you are unsure how to answer this question, answer N.
 
@@ -57,7 +56,7 @@ config SECURITY_NETWORK
 config PAGE_TABLE_ISOLATION
 	bool "Remove the kernel mapping in user mode"
 	default y
-	depends on X86_64 && !UML
+	depends on (X86_64 || X86_PAE) && !UML
 	help
 	  This feature reduces the number of hardware side channels by
 	  ensuring that the majority of kernel addresses are not mapped
@@ -153,7 +152,6 @@ config HAVE_HARDENED_USERCOPY_ALLOCATOR
 config HARDENED_USERCOPY
 	bool "Harden memory copies between kernel and userspace"
 	depends on HAVE_HARDENED_USERCOPY_ALLOCATOR
-	select BUG
 	imply STRICT_DEVMEM
 	help
 	  This option checks for obviously wrong memory regions when
@@ -231,20 +229,21 @@ config STATIC_USERMODEHELPER_PATH
 	  If you wish for all usermode helper programs to be disabled,
 	  specify an empty string here (i.e. "").
 
-source security/selinux/Kconfig
-source security/smack/Kconfig
-source security/tomoyo/Kconfig
-source security/apparmor/Kconfig
-source security/loadpin/Kconfig
-source security/yama/Kconfig
-source security/can-exec/Kconfig
-source security/hashcheck/Kconfig
-source security/whitelist/Kconfig
+source "security/selinux/Kconfig"
+source "security/smack/Kconfig"
+source "security/tomoyo/Kconfig"
+source "security/apparmor/Kconfig"
+source "security/loadpin/Kconfig"
+source "security/yama/Kconfig"
+source "security/safesetid/Kconfig"
+source "security/can-exec/Kconfig"
+source "security/hashcheck/Kconfig"
+source "security/whitelist/Kconfig"
 
-source security/integrity/Kconfig
+source "security/integrity/Kconfig"
 
 choice
-	prompt "Default security module"
+	prompt "First legacy 'major LSM' to be initialized"
 	default DEFAULT_SECURITY_SELINUX if SECURITY_SELINUX
 	default DEFAULT_SECURITY_SMACK if SECURITY_SMACK
 	default DEFAULT_SECURITY_TOMOYO if SECURITY_TOMOYO
@@ -255,8 +254,13 @@ choice
 	default DEFAULT_SECURITY_DAC
 
 	help
-	  Select the security module that will be used by default if the
-	  kernel parameter security= is not specified.
+	  This choice is there only for converting CONFIG_DEFAULT_SECURITY
+	  in old kernel configs to CONFIG_LSM in new kernel configs. Don't
+	  change this choice unless you are creating a fresh kernel config,
+	  for this choice will be ignored after CONFIG_LSM has been set.
+
+	  Selects the legacy "major security module" that will be
+	  initialized first. Overridden by non-default CONFIG_LSM.
 
 	config DEFAULT_SECURITY_SELINUX
 		bool "SELinux" if SECURITY_SELINUX=y
@@ -284,16 +288,20 @@ choice
 
 endchoice
 
-config DEFAULT_SECURITY
-	string
-	default "selinux" if DEFAULT_SECURITY_SELINUX
-	default "smack" if DEFAULT_SECURITY_SMACK
-	default "tomoyo" if DEFAULT_SECURITY_TOMOYO
-	default "apparmor" if DEFAULT_SECURITY_APPARMOR
-	default "can-exec" if DEFAULT_SECURITY_CAN_EXEC
-	default "hashcheck" if DEFAULT_SECURITY_HASH_CHECK
-	default "whitelist" if DEFAULT_SECURITY_WHITELIST
-	default "" if DEFAULT_SECURITY_DAC
+config LSM
+	string "Ordered list of enabled LSMs"
+	default "yama,loadpin,safesetid,integrity,smack,selinux,tomoyo,apparmor" if DEFAULT_SECURITY_SMACK
+	default "yama,loadpin,safesetid,integrity,apparmor,selinux,smack,tomoyo" if DEFAULT_SECURITY_APPARMOR
+	default "yama,loadpin,safesetid,integrity,tomoyo" if DEFAULT_SECURITY_TOMOYO
+	default "yama,loadpin,safesetid,integrity,can-exec" if DEFAULT_SECURITY_CAN_EXEC
+	default "yama,loadpin,safesetid,integrity" if DEFAULT_SECURITY_DAC
+	default "yama,loadpin,safesetid,integrity,can-exec,selinux,smack,tomoyo,apparmor"
+	help
+	  A comma-separated list of LSMs, in initialization order.
+	  Any LSMs left off this list will be ignored. This can be
+	  controlled at boot with the "lsm=" parameter.
+
+	  If unsure, leave this as the default.
 
 endmenu
 

--- a/security/Makefile
+++ b/security/Makefile
@@ -1,3 +1,4 @@
+# SPDX-License-Identifier: GPL-2.0
 #
 # Makefile for the kernel security code
 #
@@ -9,6 +10,7 @@ subdir-$(CONFIG_SECURITY_TOMOYO)        += tomoyo
 subdir-$(CONFIG_SECURITY_APPARMOR)	+= apparmor
 subdir-$(CONFIG_SECURITY_YAMA)		+= yama
 subdir-$(CONFIG_SECURITY_LOADPIN)	+= loadpin
+subdir-$(CONFIG_SECURITY_SAFESETID)    += safesetid
 subdir-$(CONFIG_SECURITY_CAN_EXEC)	+= can-exec
 subdir-$(CONFIG_SECURITY_HASH_CHECK)	+= hashcheck
 subdir-$(CONFIG_SECURITY_WHITELIST)	+= whitelist
@@ -27,6 +29,7 @@ obj-$(CONFIG_SECURITY_TOMOYO)		+= tomoyo/
 obj-$(CONFIG_SECURITY_APPARMOR)		+= apparmor/
 obj-$(CONFIG_SECURITY_YAMA)		+= yama/
 obj-$(CONFIG_SECURITY_LOADPIN)		+= loadpin/
+obj-$(CONFIG_SECURITY_SAFESETID)       += safesetid/
 obj-$(CONFIG_CGROUP_DEVICE)		+= device_cgroup.o
 obj-$(CONFIG_SECURITY_CAN_EXEC)		+= can-exec/
 obj-$(CONFIG_SECURITY_HASH_CHECK)	+= hashcheck/
@@ -35,3 +38,4 @@ obj-$(CONFIG_SECURITY_WHITELIST)	+= whitelist/
 # Object integrity file lists
 subdir-$(CONFIG_INTEGRITY)		+= integrity
 obj-$(CONFIG_INTEGRITY)			+= integrity/
+

--- a/security/can-exec/README.md
+++ b/security/can-exec/README.md
@@ -4,11 +4,11 @@ can-exec
 This is a LSM in which the kernel calls a user-mode helper to decide
 if binaries should be executed.
 
-Every time a command is to be executed the kernel will invoke:
+Every time a command is to be executed the kernel will invoke a user-space helper:
 
     /sbin/can-exec $UID $COMMAND
 
-Where the arguments are the UID of the invoking user, and the command
+The arguments supplied are the UID of the invoking user, and the command
 they're trying to execute.  If the user-space binary exits with a return-code
 of zero the execution will be permitted, otherwise it will be denied.
 
@@ -51,6 +51,21 @@ Once the user-space binary is in-place you'll need to enable the enforcement
 by running:
 
      echo 1 > /proc/sys/kernel/can-exec/enabled
+
+
+Testing
+-------
+
+To test that the module is enabled you'll want to run:
+
+    $ echo $(cat /sys/kernel/security/lsm)
+    capability,safesetid,can_exec
+
+If you see `can_exec` listed, and you get output from this command you're good:
+
+    $ dmesg | grep LSM
+    [    0.282365] LSM: Security Framework initializing
+    [    0.283323] LSM initialized: can_exec
 
 
 Links

--- a/security/can-exec/README.md
+++ b/security/can-exec/README.md
@@ -17,8 +17,53 @@ of zero the execution will be permitted, otherwise it will be denied.
 Installation & Configuration
 ----------------------------
 
-Build the kernel with this support enabled, and then you'll need to
-configure the user-space side.
+First of all you'll need to build the kernel, with this module enabled.  Since stacking exists these days here are the settings I used to test:
+
+     #
+     # Security options
+     #
+     CONFIG_KEYS=y
+     CONFIG_KEYS_COMPAT=y
+     # CONFIG_PERSISTENT_KEYRINGS is not set
+     # CONFIG_BIG_KEYS is not set
+     # CONFIG_TRUSTED_KEYS is not set
+     CONFIG_ENCRYPTED_KEYS=m
+     # CONFIG_KEY_DH_OPERATIONS is not set
+     CONFIG_SECURITY_DMESG_RESTRICT=y
+     CONFIG_SECURITY=y
+     CONFIG_SECURITYFS=y
+     CONFIG_SECURITY_NETWORK=y
+     CONFIG_PAGE_TABLE_ISOLATION=y
+     CONFIG_SECURITY_NETWORK_XFRM=y
+     CONFIG_SECURITY_PATH=y
+     # CONFIG_INTEL_TXT is not set
+     CONFIG_HAVE_HARDENED_USERCOPY_ALLOCATOR=y
+     # CONFIG_HARDENED_USERCOPY is not set
+     # CONFIG_FORTIFY_SOURCE is not set
+     # CONFIG_STATIC_USERMODEHELPER is not set
+     # CONFIG_SECURITY_SELINUX is not set
+     # CONFIG_SECURITY_SMACK is not set
+     # CONFIG_SECURITY_TOMOYO is not set
+     # CONFIG_SECURITY_APPARMOR is not set
+     # CONFIG_SECURITY_LOADPIN is not set
+     # CONFIG_SECURITY_YAMA is not set
+     CONFIG_SECURITY_SAFESETID=y
+     CONFIG_SECURITY_CAN_EXEC=y
+     # CONFIG_SECURITY_HASH_CHECK is not set
+     # CONFIG_SECURITY_WHITELIST is not set
+     # CONFIG_INTEGRITY is not set
+     CONFIG_DEFAULT_SECURITY_CAN_EXEC=y
+     # CONFIG_DEFAULT_SECURITY_DAC is not set
+     CONFIG_LSM="yama,loadpin,safesetid,integrity,can-exec,selinux,smack,tomoyo,apparmor"
+     CONFIG_XOR_BLOCKS=m
+     CONFIG_ASYNC_CORE=m
+     CONFIG_ASYNC_MEMCPY=m
+     CONFIG_ASYNC_XOR=m
+     CONFIG_ASYNC_PQ=m
+     CONFIG_ASYNC_RAID6_RECOV=m
+     CONFIG_CRYPTO=y
+
+Once you've rebooted into your new kernel, and confirmed normal operation is good, you'll then need to configure the user-space side.
 
 * Install `/sbin/can-exec` from the `samples/` directory.
    * This will be invoked to decide if users can execute binaries.

--- a/security/can-exec/can_exec_lsm.c
+++ b/security/can-exec/can_exec_lsm.c
@@ -236,4 +236,11 @@ static int __init can_exec_init(void)
     return 0;
 }
 
-security_initcall(can_exec_init);
+
+/*
+ * Ensure the initialization code is called.
+ */
+DEFINE_LSM(can_exec_init) = {
+        .init = can_exec_init,
+        .name = "can-exec",
+};

--- a/security/hashcheck/hashcheck_lsm.c
+++ b/security/hashcheck/hashcheck_lsm.c
@@ -295,4 +295,11 @@ static int __init hashcheck_init(void)
     return 0;
 }
 
-security_initcall(hashcheck_init);
+
+/*
+ * Ensure the initialization code is called.
+ */
+DEFINE_LSM(hashcheck_init) = {
+        .init = hashcheck_init,
+        .name = "hashcheck",
+};

